File: packages/ios/Sources/MushiMushi/Mushi.swift
import Foundation
import Network
#if os(iOS)
import UIKit
#endif

/// Public entry point for the iOS SDK. Mirrors the surface of
/// `@mushi-mushi/core` `Mushi` so cross-platform docs stay accurate.
///
/// Usage:
/// ```swift
/// import MushiMushi
///
/// @main struct MyApp: App {
///   init() {
///     Mushi.shared.configure(with: MushiConfig(projectId: "...", apiKey: "..."))
///   }
///   var body: some Scene { ... }
/// }
/// ```
public final class Mushi {
    public static let shared = Mushi()

    // All `_*` state is read/written under `stateLock`. Mirrors the
    // `synchronized(this)` pattern in the Android `Mushi` object so
    // a host that calls `configure(...)` from a non-main thread while
    // a background `report(...)` is in flight can't observe a torn
    // snapshot (e.g. new `apiClient` paired with old `config`).
    private let stateLock = NSLock()
    private var _config: MushiConfig?
    private var _apiClient: ApiClient?
    private var _queue: OfflineQueue?
    private var _flushTimer: Timer?
    private var _user: [String: Any]?
    private var _globalMetadata: [String: Any] = [:]
    private var _breadcrumbs = BreadcrumbCollector()
    private var _piiScrubber = PIIScrubber()
    // Added: network-aware delivery (Phase 2.4)
    private var _networkMonitor: NWPathMonitor?
    private let networkMonitorQueue = DispatchQueue(label: "mushi.network-monitor")
    #if os(iOS)
    private var _proactiveDetector: ProactiveDetector?
    private weak var _floatingButton: UIButton?
    private var _foregroundObserver: NSObjectProtocol?
    private var _proactiveForegroundObserver: NSObjectProtocol?
    private var _proactiveKeyWindowObserver: NSObjectProtocol?
    #endif

    private init() {}

    // MARK: - Locked accessors
    //
    // Trivial single-property getters that take the lock for the read.
    // Reads that need to span multiple properties go through `snapshot()`
    // below so the bundle is consistent.
    private func withLock<T>(_ body: () -> T) -> T {
        stateLock.lock()
        defer { stateLock.unlock() }
        return body()
    }

    private struct ReportSnapshot {
        let config: MushiConfig
        let apiClient: ApiClient
        let breadcrumbs: BreadcrumbCollector
        let piiScrubber: PIIScrubber
        let mergedMetadata: [String: Any]
    }

    /// Snapshot every mutable property `report(...)` reads, in one critical
    /// section. Merging metadata also happens under the lock so `setUser` /
    /// `setMetadata` / `configure` mid-report can't slip in a half-applied
    /// edit.
    private func reportSnapshot(extraMetadata: [String: Any]?) -> ReportSnapshot? {
        withLock {
            guard let config = _config, let apiClient = _apiClient else { return nil }
            var merged = _globalMetadata
            extraMetadata?.forEach { merged[$0.key] = $0.value }
            if let user = _user { merged["user"] = user }
            return ReportSnapshot(
                config: config,
                apiClient: apiClient,
                breadcrumbs: _breadcrumbs,
                piiScrubber: _piiScrubber,
                mergedMetadata: merged
            )
        }
    }

    public func configure(with config: MushiConfig) {
        // Phase 1: take the lock and apply every reassignment atomically.
        // Side-effecting setup (timer / network monitor / triggers) runs
        // *outside* the lock so the host's notification observers — which
        // may call back into `addBreadcrumb` — never deadlock.
        let breadcrumbsRef = withLock { () -> BreadcrumbCollector in
            self._config = config
            self._piiScrubber = PIIScrubber(config: config.pii)
            if config.captureBreadcrumbs {
                self._breadcrumbs = BreadcrumbCollector()
            }
            let q = OfflineQueue(maxBytes: config.offlineQueueMaxBytes)
            self._queue = q
            self._apiClient = ApiClient(config: config, queue: q)
            return self._breadcrumbs
        }
        #if os(iOS)
        // Tear down any prior session — these operations take the lock
        // internally where they need to read the current detector / observer.
        teardownProactiveSession()
        removeFloatingButton()
        clearForegroundObservers()
        #endif
        installTriggers()
        startFlushTimer()
        startNetworkMonitor()
        breadcrumbsRef.add(category: .lifecycle, message: "Mushi configured")
    }

    /// Append a breadcrumb to the ring buffer. Included automatically on the
    /// next `report()` call. Mirrors `Mushi.addBreadcrumb()` in the web SDK.
    public func addBreadcrumb(
        category: MushiBreadcrumb.Category,
        level: MushiBreadcrumb.Level = .info,
        message: String,
        data: [String: String]? = nil
    ) {
        withLock { _breadcrumbs }.add(
            category: category, level: level, message: message, data: data
        )
    }

    /// Snapshot of the current breadcrumb ring buffer, oldest first.
    public func getBreadcrumbs() -> [MushiBreadcrumb] {
        withLock { _breadcrumbs }.getAll()
    }

    /// Attach app/user identity to subsequent native reports.
    public func setUser(_ user: [String: Any]?) {
        withLock { _user = user }
    }

    /// Attach or clear a metadata key that should be sent with subsequent
    /// native reports.
    public func setMetadata(_ key: String, value: Any?) {
        withLock {
            if let value {
                _globalMetadata[key] = value
            } else {
                _globalMetadata.removeValue(forKey: key)
            }
        }
    }

    #if os(iOS)
    public func setHidden(_ hidden: Bool) {
        hidden ? removeFloatingButton() : installFloatingButtonIfNeeded()
    }

    public func attachTo(_ control: UIControl) {
        control.addTarget(self, action: #selector(showWidgetFromButton), for: .touchUpInside)
    }
    #endif

    /// Submit a report with the given description and optional category.
    /// Captures device context and (if enabled) a screenshot automatically.
    /// Attaches the breadcrumb ring buffer (PII-scrubbed) to the payload.
    public func report(
        description: String,
        category: String = "bug",
        metadata: [String: Any]? = nil
    ) {
        guard let snapshot = reportSnapshot(extraMetadata: metadata) else { return }

        let scrubbedDescription = snapshot.piiScrubber.scrub(description)
        var payload: [String: Any] = [
            "description": scrubbedDescription,
            "category": category,
            "context": DeviceContext.capture()
        ]
        if !snapshot.mergedMetadata.isEmpty { payload["metadata"] = snapshot.mergedMetadata }

        if snapshot.config.captureBreadcrumbs {
            let crumbs = snapshot.breadcrumbs.getAll().map { crumb -> [String: Any] in
                var d: [String: Any] = [
                    "timestamp": crumb.timestamp,
                    "category": crumb.category.rawValue,
                    "level": crumb.level.rawValue,
                    "message": snapshot.piiScrubber.scrub(crumb.message),
                ]
                if let data = crumb.data {
                    d["data"] = data.mapValues { snapshot.piiScrubber.scrub($0) }
                }
                return d
            }
            if !crumbs.isEmpty { payload["breadcrumbs"] = crumbs }
        }

        snapshot.breadcrumbs.add(category: .lifecycle, message: "report submitted: \(category)")

        #if os(iOS)
        if snapshot.config.captureScreenshot {
            let submit: ([String: Any]) -> Void = { snapshot.apiClient.submitReport($0) }
            if Thread.isMainThread {
                MainActor.assumeIsolated {
                    var p = payload
                    if let s = ScreenshotCapture.captureBase64() {
                        p["screenshot"] = s
                    }
                    submit(p)
                }
            } else {
                DispatchQueue.main.async {
                    MainActor.assumeIsolated {
                        var p = payload
                        if let s = ScreenshotCapture.captureBase64() {
                            p["screenshot"] = s
                        }
                        submit(p)
                    }
                }
            }
            return
        }
        #endif
        snapshot.apiClient.submitReport(payload)
    }

    /// Capture a Swift `Error`. Normalises the error (name/message/stack/cause)
    /// and files a report with structured metadata — mirrors `captureException`
    /// in the web SDK.
    public func captureError(_ error: Error, context: [String: Any]? = nil) {
        let norm = normaliseError(error)
        var meta = context ?? [:]
        meta["error"] = normaliseExceptionToMetadata(norm)
        report(
            description: norm.message,
            category: "bug",
            metadata: meta
        )
    }

    /// Programmatically present the bottom sheet widget.
    public func showWidget(category: String? = nil, metadata: [String: Any]? = nil) {
        #if os(iOS)
        // Snapshot config + apiClient atomically so a concurrent reconfigure
        // can't pair the old config with a new client (or vice versa) when
        // the user finally taps "Submit" inside the widget.
        let snapshot: (config: MushiConfig, apiClient: ApiClient)? = withLock {
            guard let c = _config, let a = _apiClient else { return nil }
            return (config: c, apiClient: a)
        }
        guard let snapshot else { return }
        DispatchQueue.main.async { @MainActor in
            guard let topVC = Self.topViewController() else { return }
            let screenshot = snapshot.config.captureScreenshot
                ? ScreenshotCapture.captureBase64()
                : nil
            let widget = MushiWidgetController(config: snapshot.config, screenshot: screenshot, initialCategory: category) { [weak self] payload in
                var report = payload
                let mergedMetadata = self?.mergeMetadata(metadata) ?? [:]
                if !mergedMetadata.isEmpty { report["metadata"] = mergedMetadata }
                snapshot.apiClient.submitReport(report)
            }
            let nav = UINavigationController(rootViewController: widget)
            topVC.present(nav, animated: true)
        }
        #endif
    }

    /// Trigger an immediate offline queue flush. Reports the number of
    /// successfully delivered items via the optional completion handler.
    public func flushOfflineQueueNow(completion: ((Int) -> Void)? = nil) {
        let client = withLock { _apiClient }
        DispatchQueue.global().async {
            client?.flushQueue()
            completion?(0)
        }
    }

    // MARK: - Private

    private func installTriggers() {
        #if os(iOS)
        guard let config = withLock({ _config }) else { return }
        ShakeDetector.onShake = nil
        if config.triggerMode == .shake || config.triggerMode == .both {
            ShakeDetector.install()
            ShakeDetector.onShake = { [weak self] in self?.showWidget() }
        }
        if config.triggerMode == .button || config.triggerMode == .both {
            let observer = NotificationCenter.default.addObserver(
                forName: UIApplication.didBecomeActiveNotification,
                object: nil,
                queue: .main
            ) { [weak self] _ in self?.installFloatingButtonIfNeeded() }
            withLock { _foregroundObserver = observer }
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) { [weak self] in
                self?.installFloatingButtonIfNeeded()
            }
        } else {
            removeFloatingButton()
        }

        // Proactive triggers (rage-tap, slow-screen)
        let proactiveConfig = ProactiveDetector.Config(
            rageTap: config.proactive.rageTap,
            slowScreen: config.proactive.slowScreen,
            slowScreenThresholdMs: config.proactive.slowScreenThresholdMs,
            maxPerSession: config.proactive.maxPerSession
        )
        let detector = ProactiveDetector(config: proactiveConfig)
        // Capture the breadcrumb collector reference now so the proactive
        // callback hot path doesn't have to take `stateLock` per event. The
        // collector itself is independently thread-safe.
        let crumbsRef = withLock { () -> BreadcrumbCollector in
            self._proactiveDetector = detector
            return self._breadcrumbs
        }

        // Reset the slow-screen frame clock when the app foregrounds, so
        // the multi-second pause while backgrounded doesn't fire as a
        // false-positive slow_screen on the first post-resume frame.
        let foregroundObs = NotificationCenter.default.addObserver(
            forName: UIApplication.willEnterForegroundNotification,
            object: nil,
            queue: .main
        ) { [weak detector] _ in detector?.resetFrameClock() }
        withLock { _proactiveForegroundObserver = foregroundObs }

        // Install on the first key window we can find. Try at +0.5 / +2 / +5 s
        // (covers slow splash + multi-window scene promotion); also listen for
        // the first didBecomeKey notification as a belt-and-braces fallback.
        let installIfPossible: () -> Bool = { [weak self, weak detector] in
            guard let self, let detector else { return true }
            guard let window = UIApplication.shared.connectedScenes
                .compactMap({ $0 as? UIWindowScene })
                .flatMap(\.windows)
                .first(where: \.isKeyWindow) else { return false }
            detector.install(in: window) { [weak self, weak crumbsRef] type, context in
                crumbsRef?.add(category: .lifecycle, level: .warning, message: "proactive:\(type)")
                self?.showWidget(category: "bug", metadata: ["proactiveTrigger": type, "proactiveContext": context])
            }
            return true
        }
        for delay in [0.5, 2.0, 5.0] {
            DispatchQueue.main.asyncAfter(deadline: .now() + delay) { _ = installIfPossible() }
        }
        let keyWindowObs = NotificationCenter.default.addObserver(
            forName: UIWindow.didBecomeKeyNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            guard let self else { return }
            if installIfPossible() {
                let toRemove = self.withLock { () -> NSObjectProtocol? in
                    let obs = self._proactiveKeyWindowObserver
                    self._proactiveKeyWindowObserver = nil
                    return obs
                }
                if let toRemove { NotificationCenter.default.removeObserver(toRemove) }
            }
        }
        withLock { _proactiveKeyWindowObserver = keyWindowObs }
        #endif
    }

    /// Build the metadata bundle for `showWidget`'s submit callback. Read
    /// lock-protected so a concurrent `setUser` / `setMetadata` mid-submit
    /// doesn't tear the merged map.
    private func mergeMetadata(_ metadata: [String: Any]?) -> [String: Any] {
        withLock {
            var merged = _globalMetadata
            metadata?.forEach { merged[$0.key] = $0.value }
            if let user = _user { merged["user"] = user }
            return merged
        }
    }

    #if os(iOS)
    /// Tear down the previous proactive detector. Called from `configure`
    /// to make reconfigure idempotent.
    private func teardownProactiveSession() {
        let prior = withLock { () -> ProactiveDetector? in
            let d = _proactiveDetector
            _proactiveDetector = nil
            return d
        }
        prior?.destroy()
    }

    /// Detach the foreground / key-window / proactive-foreground notification
    /// observers (if any). Safe to call from `configure` repeatedly.
    private func clearForegroundObservers() {
        let observers: [NSObjectProtocol] = withLock {
            let bundle: [NSObjectProtocol?] = [
                _foregroundObserver,
                _proactiveForegroundObserver,
                _proactiveKeyWindowObserver,
            ]
            _foregroundObserver = nil
            _proactiveForegroundObserver = nil
            _proactiveKeyWindowObserver = nil
            return bundle.compactMap { $0 }
        }
        observers.forEach { NotificationCenter.default.removeObserver($0) }
    }
    #endif

    #if os(iOS)
    private func installFloatingButtonIfNeeded() {
        // UIKit views are main-thread only, but we still need atomic reads
        // of `_config` + `_floatingButton` so a concurrent reconfigure
        // doesn't leave us with a button bound to the previous theme.
        guard let config = withLock({ _config }) else { return }
        guard withLock({ _floatingButton }) == nil else { return }
        guard config.triggerMode == .button || config.triggerMode == .both else { return }
        guard let window = UIApplication.shared.connectedScenes
            .compactMap({ $0 as? UIWindowScene })
            .flatMap(\.windows)
            .first(where: \.isKeyWindow) else { return }

        let button = UIButton(type: .system)
        button.setTitle("🐛", for: .normal)
        button.titleLabel?.font = .systemFont(ofSize: 24)
        button.accessibilityLabel = "Report a bug"
        button.backgroundColor = config.theme.dark ? .black : .systemBackground
        button.tintColor = UIColor(hex: config.theme.accentColor) ?? .systemBlue
        button.layer.cornerRadius = 28
        button.layer.borderWidth = 1
        button.layer.borderColor = button.tintColor.cgColor
        button.layer.shadowColor = UIColor.black.cgColor
        button.layer.shadowOpacity = 0.18
        button.layer.shadowRadius = 12
        button.layer.shadowOffset = CGSize(width: 0, height: 6)
        button.translatesAutoresizingMaskIntoConstraints = false
        button.addTarget(self, action: #selector(showWidgetFromButton), for: .touchUpInside)
        window.addSubview(button)
        let inset = config.triggerInset
        var constraints = [
            button.widthAnchor.constraint(equalToConstant: 56),
            button.heightAnchor.constraint(equalToConstant: 56),
            button.bottomAnchor.constraint(equalTo: window.safeAreaLayoutGuide.bottomAnchor, constant: -inset.bottom)
        ]
        if let leading = inset.leading {
            constraints.append(button.leadingAnchor.constraint(equalTo: window.safeAreaLayoutGuide.leadingAnchor, constant: leading))
        } else {
            constraints.append(button.trailingAnchor.constraint(equalTo: window.safeAreaLayoutGuide.trailingAnchor, constant: -(inset.trailing ?? 20)))
        }
        NSLayoutConstraint.activate(constraints)
        withLock { _floatingButton = button }
    }

    @objc private func showWidgetFromButton() {
        showWidget()
    }

    private func removeFloatingButton() {
        let prior = withLock { () -> UIButton? in
            let b = _floatingButton
            _floatingButton = nil
            return b
        }
        prior?.removeFromSuperview()
    }
    #endif

    private func startFlushTimer() {
        let oldTimer = withLock { () -> Timer? in
            let t = _flushTimer
            _flushTimer = nil
            return t
        }
        oldTimer?.invalidate()
        let timer = Timer.scheduledTimer(withTimeInterval: 30, repeats: true) { [weak self] _ in
            self?.withLock { self?._apiClient }?.flushQueue()
        }
        withLock { _flushTimer = timer }
        // Also try once immediately, in case the app was opened with stale queued reports.
        let client = withLock { _apiClient }
        DispatchQueue.global().async { client?.flushQueue() }
    }

    // Added: network-aware delivery (Phase 2.4)
    private func startNetworkMonitor() {
        let oldMonitor = withLock { () -> NWPathMonitor? in
            let m = _networkMonitor
            _networkMonitor = nil
            return m
        }
        oldMonitor?.cancel()
        let monitor = NWPathMonitor()
        monitor.pathUpdateHandler = { [weak self] path in
            guard path.status == .satisfied else { return }
            // `withLock` returns `ApiClient?`; through a weak `self?.` it
            // becomes `ApiClient??` so flatten before scheduling the flush.
            let client: ApiClient? = self?.withLock { self?._apiClient } ?? nil
            DispatchQueue.global().async { client?.flushQueue() }
        }
        // Reuse a single monitor queue across reconfigure calls — no per-call
        // queue allocation.
        monitor.start(queue: networkMonitorQueue)
        withLock { _networkMonitor = monitor }
    }

    #if os(iOS)
    private static func topViewController(base: UIViewController? = nil) -> UIViewController? {
        let root = base ?? UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .flatMap(\.windows)
            .first(where: \.isKeyWindow)?.rootViewController

        if let nav = root as? UINavigationController { return topViewController(base: nav.visibleViewController) }
        if let tab = root as? UITabBarController { return topViewController(base: tab.selectedViewController) }
        if let presented = root?.presentedViewController { return topViewController(base: presented) }
        return root
    }
    #endif
}

File: packages/capacitor/android/mushi-android/src/main/kotlin/dev/mushimushi/sdk/Mushi.kt
package dev.mushimushi.sdk

import android.app.Activity
import android.app.Application
import android.content.Context
import android.graphics.Color
import android.graphics.drawable.GradientDrawable
import android.content.Intent
import android.net.ConnectivityManager
import android.net.Network
import android.os.Bundle
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.TextView
import androidx.fragment.app.FragmentActivity
import dev.mushimushi.sdk.capture.BreadcrumbCollector
import dev.mushimushi.sdk.capture.DeviceContext
import dev.mushimushi.sdk.capture.MushiBreadcrumb
import dev.mushimushi.sdk.capture.ScreenshotCapture
import dev.mushimushi.sdk.capture.ShakeDetector
import dev.mushimushi.sdk.capture.ExceptionNormaliser
import dev.mushimushi.sdk.capture.normaliseThrowable
import dev.mushimushi.sdk.capture.ProactiveDetector
import dev.mushimushi.sdk.config.MushiConfig
import dev.mushimushi.sdk.config.TriggerMode
import dev.mushimushi.sdk.network.ApiClient
import dev.mushimushi.sdk.pii.PIIScrubber
import dev.mushimushi.sdk.pii.PIIScrubberConfig
import dev.mushimushi.sdk.storage.OfflineQueue
import dev.mushimushi.sdk.widget.MushiBottomSheet
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit

/**
 * Public entry point for the Android SDK. Mirrors the surface of the iOS
 * `Mushi` and the JS `@mushi-mushi/core` SDK so cross-platform docs stay
 * accurate. Thread-safe; safe to call from any thread.
 */
object Mushi {
    @Volatile private var initialized = false
    @Volatile private var config: MushiConfig? = null
    private var queue: OfflineQueue? = null
    private var apiClient: ApiClient? = null
    private var shakeDetector: ShakeDetector? = null
    private var proactiveDetector: ProactiveDetector? = null
    private var breadcrumbs = BreadcrumbCollector()
    private var piiScrubber = PIIScrubber()
    private var currentActivity: Activity? = null
    private var floatingButton: View? = null
    private var floatingButtonOwner: Activity? = null
    private var flushExecutor: ScheduledExecutorService? = null
    private var lifecycleRegistered = false
    private var user: Map<String, Any?>? = null
    private val globalMetadata = mutableMapOf<String, Any?>()

    /** Listener fired after every successful report submission. Used by the
     *  optional Sentry bridge to mirror reports into Sentry's UserFeedback. */
    @Volatile var onReportSubmitted: ((Map<String, Any?>) -> Unit)? = null

    fun init(app: Application, config: MushiConfig) {
        synchronized(this) {
            this.config = config
            this.piiScrubber = PIIScrubber(PIIScrubberConfig(
                emails = config.pii.emails,
                phones = config.pii.phones,
                creditCards = config.pii.creditCards,
                ssns = config.pii.ssns,
                ipAddresses = config.pii.ipAddresses,
                secretTokens = config.pii.secretTokens,
                ipv6 = config.pii.ipv6,
            ))
            if (config.captureBreadcrumbs) breadcrumbs = BreadcrumbCollector()
            val q = queue ?: OfflineQueue(app, config.offlineQueueMaxBytes)
            this.queue = q
            this.apiClient = ApiClient(config, q) { payload ->
                onReportSubmitted?.invoke(payload)
            }

            if (!lifecycleRegistered) {
                app.registerActivityLifecycleCallbacks(LifecycleTracker())
                lifecycleRegistered = true
            }
            refreshTriggers(app)
            refreshFloatingButton()
            if (!initialized) {
                startFlushTimer()
                // Added: network-aware delivery (Phase 2.4)
                startNetworkMonitor(app)
            }
            initialized = true
            breadcrumbs.add(MushiBreadcrumb.Category.LIFECYCLE, message = "Mushi configured")
        }
    }

    fun setUser(user: Map<String, Any?>?) {
        synchronized(this) { this.user = user }
    }

    fun setMetadata(key: String, value: Any?) {
        synchronized(this) {
            if (value == null) globalMetadata.remove(key) else globalMetadata[key] = value
        }
    }

    /** Append a breadcrumb to the ring buffer. Mirrors `Mushi.addBreadcrumb()` in the web SDK. */
    fun addBreadcrumb(
        category: MushiBreadcrumb.Category,
        level: MushiBreadcrumb.Level = MushiBreadcrumb.Level.INFO,
        message: String,
        data: Map<String, String>? = null,
    ) = breadcrumbs.add(category, level, message, data)

    /** Snapshot of the current breadcrumb ring buffer, oldest first. */
    fun getBreadcrumbs(): List<MushiBreadcrumb> = breadcrumbs.getAll()

    fun setHidden(hidden: Boolean) {
        if (hidden) removeFloatingButton() else refreshFloatingButton()
    }

    fun attachTo(view: View) {
        view.setOnClickListener { showWidget() }
    }

    /** Submit a report. Auto-attaches device context, breadcrumbs (PII-scrubbed), and screenshot. */
    fun report(
        description: String,
        category: String = "bug",
        metadata: Map<String, Any?>? = null
    ) {
        val cfg = config ?: return
        val client = apiClient ?: return
        val ctx = currentActivity?.applicationContext ?: return

        val scrubbedDescription = piiScrubber.scrub(description)
        val payload = mutableMapOf<String, Any?>(
            "description" to scrubbedDescription,
            "category" to category,
            "context" to DeviceContext.capture(ctx)
        )
        mergeMetadata(metadata).takeIf { it.isNotEmpty() }?.let { payload["metadata"] = it }

        if (cfg.captureBreadcrumbs) {
            val crumbs = breadcrumbs.getAll().map { crumb ->
                crumb.toMap().toMutableMap().apply {
                    put("message", piiScrubber.scrub(crumb.message))
                    crumb.data?.let { put("data", it.mapValues { (_, v) -> piiScrubber.scrub(v) }) }
                }
            }
            if (crumbs.isNotEmpty()) payload["breadcrumbs"] = crumbs
        }

        breadcrumbs.add(MushiBreadcrumb.Category.LIFECYCLE, message = "report submitted: $category")

        if (cfg.captureScreenshot) {
            currentActivity?.let { act ->
                ScreenshotCapture.captureBase64(act)?.let { payload["screenshot"] = it }
            }
        }
        client.submitReport(payload)
    }

    fun captureError(throwable: Throwable, metadata: Map<String, Any?>? = null) {
        val norm = normaliseThrowable(throwable)
        val meta = (metadata ?: emptyMap()).toMutableMap()
        meta["error"] = norm.toMap()
        report(description = norm.message, category = "bug", metadata = meta)
    }

    /** Programmatically present the bottom-sheet widget. */
    fun showWidget(category: String? = null, metadata: Map<String, Any?>? = null) {
        val cfg = config ?: return
        val act = currentActivity as? FragmentActivity ?: return
        val screenshot = if (cfg.captureScreenshot) ScreenshotCapture.captureBase64(act) else null
        MushiBottomSheet().apply {
            this.config = cfg
            this.attachedScreenshot = screenshot
            this.initialCategory = category
            this.onSubmit = { payload ->
                val full = payload.toMutableMap().apply {
                    put("context", DeviceContext.capture(act.applicationContext))
                    mergeMetadata(metadata).takeIf { it.isNotEmpty() }?.let { put("metadata", it) }
                }
                apiClient?.submitReport(full)
            }
        }.show(act.supportFragmentManager, "mushi-bottom-sheet")
    }

    private fun mergeMetadata(metadata: Map<String, Any?>?): Map<String, Any?> = synchronized(this) {
        buildMap {
            putAll(globalMetadata)
            metadata?.let { putAll(it) }
            user?.let { put("user", it) }
        }
    }

    private fun refreshTriggers(app: Application) {
        val cfg = config ?: return
        shakeDetector?.stop()
        shakeDetector = null
        if (cfg.triggerMode == TriggerMode.SHAKE || cfg.triggerMode == TriggerMode.BOTH) {
            shakeDetector = ShakeDetector(app) { showWidget() }.also { it.start() }
        }
        proactiveDetector?.destroy()
        proactiveDetector = null
        val pdConfig = ProactiveDetector.Config(
            rageTap = cfg.proactive.rageTap,
            slowScreen = cfg.proactive.slowScreen,
            slowScreenThresholdMs = cfg.proactive.slowScreenThresholdMs,
            maxPerSession = cfg.proactive.maxPerSession,
        )
        proactiveDetector = ProactiveDetector(pdConfig)
    }

    private fun refreshFloatingButton() {
        removeFloatingButton()
        currentActivity?.let { installButtonIfNeeded(it) }
    }

    private fun installButtonIfNeeded(activity: Activity) {
        val cfg = config ?: return
        if (cfg.triggerMode != TriggerMode.BUTTON && cfg.triggerMode != TriggerMode.BOTH) return
        if (floatingButton?.parent != null && floatingButtonOwner === activity) return
        removeFloatingButton()

        val root = activity.findViewById<ViewGroup>(android.R.id.content) ?: return
        val accent = runCatching { Color.parseColor(cfg.theme.accentColor) }.getOrDefault(Color.parseColor("#22C55E"))
        val button = TextView(activity).apply {
            text = "\uD83D\uDC1B"
            textSize = 22f
            gravity = Gravity.CENTER
            contentDescription = "Report a bug"
            elevation = 18f
            background = GradientDrawable().apply {
                shape = GradientDrawable.RECTANGLE
                cornerRadius = 28f
                setColor(if (cfg.theme.dark) Color.BLACK else Color.WHITE)
                setStroke(2, accent)
            }
            setOnClickListener { showWidget() }
        }
        val density = activity.resources.displayMetrics.density
        val size = (56 * density).toInt()
        val inset = cfg.triggerInset
        val marginStart = ((inset.startDp ?: 20) * density).toInt()
        val marginEnd = ((inset.endDp ?: 20) * density).toInt()
        val bottom = (inset.bottomDp * density).toInt()
        root.addView(button, FrameLayout.LayoutParams(size, size).apply {
            gravity = Gravity.BOTTOM or if (inset.startDp != null) Gravity.START else Gravity.END
            setMargins(marginStart, marginStart, marginEnd, bottom)
        })
        floatingButton = button
        floatingButtonOwner = activity
    }

    private fun removeFloatingButton() {
        (floatingButton?.parent as? ViewGroup)?.removeView(floatingButton)
        floatingButton = null
        floatingButtonOwner = null
    }

    private fun startFlushTimer() {
        flushExecutor?.shutdownNow()
        flushExecutor = Executors.newSingleThreadScheduledExecutor { r ->
            Thread(r, "mushi-flush").apply { isDaemon = true }
        }.also { ex ->
            ex.scheduleWithFixedDelay({
                runCatching { apiClient?.flushQueue() }
            }, 5, 30, TimeUnit.SECONDS)
        }
    }

    // Added: network-aware delivery (Phase 2.4)
    private fun startNetworkMonitor(app: Application) {
        val connectivityManager =
            app.getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager
                ?: return
        val networkCallback = object : ConnectivityManager.NetworkCallback() {
            override fun onAvailable(network: Network) {
                flushExecutor?.execute { runCatching { apiClient?.flushQueue() } }
            }
        }
        runCatching { connectivityManager.registerDefaultNetworkCallback(networkCallback) }
    }

    private class LifecycleTracker : Application.ActivityLifecycleCallbacks {
        override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) = Unit
        override fun onActivityStarted(activity: Activity) {
            currentActivity = activity
            installButtonIfNeeded(activity)
        }
        override fun onActivityResumed(activity: Activity) {
            currentActivity = activity
            installButtonIfNeeded(activity)
            proactiveDetector?.let { detector ->
                val window = activity.window ?: return@let
                detector.resetFrameClock()
                detector.install(window) { type, context ->
                    breadcrumbs.add(MushiBreadcrumb.Category.LIFECYCLE, MushiBreadcrumb.Level.WARNING,
                        "proactive:$type")
                    showWidget(category = "bug", metadata = mapOf(
                        "proactiveTrigger" to type,
                        "proactiveContext" to context,
                    ))
                }
            }
        }
        override fun onActivityPaused(activity: Activity) {
            if (currentActivity === activity) {
                proactiveDetector?.uninstall()
            }
        }
        override fun onActivityStopped(activity: Activity) {
            if (floatingButtonOwner === activity) removeFloatingButton()
            if (currentActivity === activity) currentActivity = null
        }
        override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) = Unit
        override fun onActivityDestroyed(activity: Activity) = Unit
    }
}
